source/distributed/fullydistributed_tria.cc

```diff
@@ -0,0 +1,562 @@
+          this->coarse_gid_to_lid.emplace_back(0, 0);
+          this->coarse_lid_to_gid.emplace_back(0, 0);
+        }
+      else
+        {
+          auto &boundary_ids = construction_data.boundary_ids;
+          auto &cells        = construction_data.cells;
+          auto &vertices     = construction_data.vertices;
+          auto &parts        = construction_data.parts;
+
+
+
+          // create inverse map
+          std::map<int, int> coarse_gid_to_lid;
+          for (auto &i : coarse_lid_to_gid)
+            coarse_gid_to_lid[i.second] = i.first;
+
+
+
+          // convert map to vector and save data structures
+          for (auto i : construction_data.coarse_lid_to_gid)
+            this->coarse_lid_to_gid.emplace_back(i);
+
+          for (auto i : coarse_gid_to_lid)
+            this->coarse_gid_to_lid.emplace_back(i);
+
+
+
+          // 1) create coarse grid
+          SubCellData subcelldata;
+          dealii::parallel::Triangulation<dim, spacedim>::create_triangulation(
+            vertices, cells, subcelldata);
+
+          for (auto c = this->begin(); c != this->end(); c++)
+            c->set_all_manifold_ids(c->material_id());
+
+
+          // 2) set boundary ids
+          int c = 0;
+          for (auto cell = this->begin_active(); cell != this->end(); ++cell)
+            for (unsigned int i = 0; i < GeometryInfo<dim>::faces_per_cell; i++)
+              {
+                unsigned int boundary_ind = boundary_ids[c++];
+                if (boundary_ind != numbers::internal_face_boundary_id)
+                  cell->face(i)->set_boundary_id(boundary_ind);
+              }
+
+          // 3) create all cell levels
+          for (unsigned int ref_counter = 1; ref_counter < parts.size();
+               ref_counter++)
+            {
+              auto coarse_cell    = this->begin(ref_counter - 1);
+              auto fine_cell_info = parts[ref_counter].begin();
+
+              for (; fine_cell_info != parts[ref_counter].end();
+                   ++fine_cell_info)
+                {
+                  auto temp = fine_cell_info->index;
+                  temp[0]   = coarse_cell_index_to_coarse_cell_id(temp[0]);
+                  while (!internal::parent<dim>(
+                    temp, coarse_cell->id().template to_binary<dim>()))
+                    coarse_cell++;
+
+                  coarse_cell->set_refine_flag();
+                }
+
+              dealii::Triangulation<dim, spacedim>::
+                execute_coarsening_and_refinement();
+            }
+
+
+
+          // 4a) set all cells artificial
+          for (auto cell = this->begin(); cell != this->end(); cell++)
+            {
+              if (cell->active())
+                cell->set_subdomain_id(
+                  dealii::numbers::artificial_subdomain_id);
+
+              if (settings & construct_multigrid_hierarchy)
+                cell->set_level_subdomain_id(
+                  dealii::numbers::artificial_subdomain_id);
```

Comment:
what level_subdomain_id do we want if ``construct_multigrid_hierarchy`` is not set? Maybe also artificial?

Reply:
Per default, I set all the `level_subdomain_id` of all cells to `dealii::numbers::artificial_subdomain_id`.

Reply:
My point here is: you should remove the if and do this call unconditionally, so that the ids are artificial if the user is not constructing the multigrid hierarchy.

Reply:
I have already removed it yesterday.